src/process_utility.c

```diff
@@ -1886,6 +1794,68 @@ process_ddl_command_end(CollectedCommand *cmd)
 	}
 }
 
+static void
+process_drop_constraint_on_chunk(Hypertable *ht, Oid chunk_relid, void *arg)
+{
+	char	   *hypertable_constraint_name = arg;
+	Chunk	   *chunk = chunk_get_by_relid(chunk_relid, ht->space->num_dimensions, true);
+
+	/* drop both metadata and table; sql_drop won't be called recursively */
+	chunk_constraint_delete_by_hypertable_constraint_name(chunk->fd.id, hypertable_constraint_name, true, true);
+}
+
+static void
+process_drop_table_constraint(EventTriggerDropTableConstraint *obj)
+{
+	/* addrnames: schema, table, constraint_name */
+
+	/* do not use relids because underlying table could be gone */
+	Hypertable *ht = hypertable_get_by_name(obj->schema, obj->table);
+
+	if (ht != NULL)
+	{
+		CatalogSecurityContext sec_ctx;
+
+		catalog_become_owner(catalog_get(), &sec_ctx);
+
+		/* Recurse to each chunk and drop the corresponding constraint */
+		foreach_chunk(ht, process_drop_constraint_on_chunk, obj->constraint_name);
+
+		catalog_restore_user(&sec_ctx);
+	}
+	else
+	{
+		Chunk	   *chunk = chunk_get_by_name(obj->schema, obj->table, 0, false);
+
+		if (NULL != chunk)
+		{
+			chunk_constraint_delete_by_constraint_name(chunk->fd.id, obj->constraint_name, true, false);
+		}
+	}
+}
+
+static void
+process_drop_index(EventTriggerDropIndex *obj)
+{
+	chunk_index_delete_by_name(obj->schema, obj->index_name, true);
+}
+
+static void
+process_ddl_sql_drop(EventTriggerDropObject *obj)
+{
+	switch (obj->type)
+	{
+		case EVENT_TRIGGER_DROP_TABLE_CONSTRAINT:
+			process_drop_table_constraint((EventTriggerDropTableConstraint *) obj);
```

Comment:
Nit: what we've done before is to cast in receiving function and `assert` the correct type.